### **Muhammad Arif Septian** <br>
### **2309106046** <br>
### **Posttest 5** <br>

#### Import Libary

In [1]:
# Import libary
import pandas as pd
import numpy as np
from sklearn.model_selection import train_test_split
from sklearn.preprocessing import StandardScaler, LabelEncoder
from tensorflow.keras.models import Sequential
from tensorflow.keras.layers import Dense, Dropout
from tensorflow.keras.optimizers import Adam

In [2]:
# Membaca dataset
data = pd.read_csv("dataset stroke clean.csv")

#### Spliting Data

In [3]:
# Membuat label encoder
le = LabelEncoder()

# Mengubah semua kolom bertipe object menjadi numerik
for col in data.columns:
    if data[col].dtype == 'object':
        data[col] = le.fit_transform(data[col].astype(str))

# Memisahkan fitur dan label
X = data.drop('stroke', axis=1)
y = data['stroke']

# Normalisasi
scaler = StandardScaler()
X = scaler.fit_transform(X)

# Split data train dan validation
X_train, X_val, y_train, y_val = train_test_split(
    X, y, test_size=0.2, random_state=42, stratify=y
)

print("Ukuran data training:", X_train.shape)
print("Ukuran data validation:", X_val.shape)



Ukuran data training: (3512, 13)
Ukuran data validation: (879, 13)


Spilting data agar siap digunakan untuk pelatihan model machine learning dengan memastikan semua nilai sudah berbentuk numerik, memiliki skala yang seimbang, dan terbagi secara proporsional antara data pelatihan serta data validasi

#### Membuat model sequential

In [4]:
# Membuat model sequential
model = Sequential([
    Dense(16, input_dim=X.shape[1], activation='relu'),
    Dropout(0.2),
    Dense(8, activation='relu'),
    Dense(1, activation='sigmoid')
])


/usr/local/lib/python3.12/dist-packages/keras/src/layers/core/dense.py:93: UserWarning: Do not pass an `input_shape`/`input_dim` argument to a layer. When using Sequential models, prefer using an `Input(shape)` object as the first layer in the model instead.
  super().__init__(activity_regularizer=activity_regularizer, **kwargs)


Membangun sebuah model jaringan saraf tiruan (neural network) berjenis Sequential yang terdiri dari beberapa lapisan terhubung secara berurutan. Model ini dirancang untuk mengenali pola dalam data dengan kombinasi lapisan padat (Dense).

#### Kompilasi Model

In [5]:
#Kompilasi model
model.compile(
    optimizer=Adam(learning_rate=0.001),
    loss='binary_crossentropy',
    metrics=['accuracy']
)

Melakukan kompilasi model dengan menentukan bagaimana model akan belajar dan dievaluasi. Optimizer Adam digunakan untuk menyesuaikan bobot secara efisien selama pelatihan, fungsi loss binary_crossentropy dipakai karena kasusnya adalah klasifikasi biner, dan metrik accuracy berfungsi untuk menilai seberapa akurat model dalam melakukan prediksi selama proses pelatihan dan evaluasi.

#### Melatih Model

In [6]:
# Melatih model
history = model.fit(
    X_train, y_train,
    epochs=50,
    batch_size=32,
    validation_data=(X_val, y_val),
    verbose=1
)

Epoch 1/50
110/110 ━━━━━━━━━━━━━━━━━━━━ 4s 5ms/step - accuracy: 0.9202 - loss: 0.4433 - val_accuracy: 0.9625 - val_loss: 0.2538
Epoch 2/50
110/110 ━━━━━━━━━━━━━━━━━━━━ 1s 3ms/step - accuracy: 0.9611 - loss: 0.2252 - val_accuracy: 0.9625 - val_loss: 0.1732
Epoch 3/50
110/110 ━━━━━━━━━━━━━━━━━━━━ 0s 3ms/step - accuracy: 0.9667 - loss: 0.1647 - val_accuracy: 0.9625 - val_loss: 0.1554
Epoch 4/50
110/110 ━━━━━━━━━━━━━━━━━━━━ 0s 3ms/step - accuracy: 0.9646 - loss: 0.1565 - val_accuracy: 0.9625 - val_loss: 0.1478
Epoch 5/50
110/110 ━━━━━━━━━━━━━━━━━━━━ 0s 3ms/step - accuracy: 0.9670 - loss: 0.1382 - val_accuracy: 0.9625 - val_loss: 0.1454
Epoch 6/50
110/110 ━━━━━━━━━━━━━━━━━━━━ 0s 3ms/step - accuracy: 0.9620 - loss: 0.1430 - val_accuracy: 0.9625 - val_loss: 0.1447
Epoch 7/50
110/110 ━━━━━━━━━━━━━━━━━━━━ 0s 3ms/step - accuracy: 0.9562 - loss: 0.1660 - val_accuracy: 0.9625 - val_loss: 0.1436
Epoch 8/50
110/110 ━━━━━━━━━━━━━━━━━━━━ 0s 3ms/step - accuracy: 0.9654 - loss: 0.1334 - val_accuracy: 0.

Tahap pelatihan model di mana jaringan saraf belajar mengenali pola dari data pelatihan dan diuji menggunakan data validasi di setiap epoch

#### Menampilkan nilai akurasi data training dan data validasi

In [7]:
# Mengambil nilai akurasi untuk data training
train_acc = history.history['accuracy'][-1]

# Mengambil nilai akurasi untuk data validasi
val_acc = history.history['val_accuracy'][-1]

# Menampilkan nilai akurasi data training dan data validasi
print(f"Akurasi Training: {train_acc*100:.2f}%")
print(f"Akurasi Validation: {val_acc*100:.2f}%")


Akurasi Training: 96.24%
Akurasi Validation: 96.25%


Berdasarkan hasil tersebut, model menunjukkan kinerja yang sangat baik dengan akurasi pelatihan sebesar 96.24% dan akurasi validasi sebesar 96.25%, yang berarti model mampu mempelajari pola data secara efektif